src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,710 @@
+          break;
+        }
+        shortlist.push(nodeData);
+      }
+      // To make the number of jumps relatively short, should connect to the node/s
+      // closest to the target first, and ask if they know of any closer nodes
+      // Then we can simply unshift the first (closest) element from the shortlist
+      shortlist.sort(function (a: NodeData, b: NodeData) {
+        if (a.distance > b.distance) {
+          return 1;
+        } else if (a.distance < b.distance) {
+          return -1;
+        } else {
+          return 0;
+        }
+      });
+    }
+    return foundAddress;
+  }
+
+  /**
+   * Performs a GRPC request to retrieve the closest nodes relative to the given
+   * target node ID.
+   * @param nodeId the node ID to search on
+   * @param targetNodeId the node ID to find other nodes closest to it
+   * @returns list of nodes and their IP/port that are closest to the target
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public async getRemoteNodeClosestNodes(
+    nodeId: NodeId,
+    targetNodeId: NodeId,
+  ): Promise<Array<NodeData>> {
+    // Construct the message
+    const nodeIdMessage = new nodesPB.Node();
+    nodeIdMessage.setNodeId(nodesUtils.encodeNodeId(targetNodeId));
+    // Send through client
+    return this.withConnF(nodeId, async (connection) => {
+      const client = await connection.getClient();
+      const response = await client.nodesClosestLocalNodesGet(nodeIdMessage);
+      const nodes: Array<NodeData> = [];
+      // Loop over each map element (from the returned response) and populate nodes
+      response.getNodeTableMap().forEach((address, nodeIdString: string) => {
+        const nodeId = nodesUtils.decodeNodeId(nodeIdString);
+        // If the nodeId is not valid we don't add it to the list of nodes
+        if (nodeId != null) {
+          nodes.push({
+            id: nodeId,
+            address: {
+              host: address.getHost() as Host | Hostname,
+              port: address.getPort() as Port,
+            },
+            distance: nodesUtils.calculateDistance(targetNodeId, nodeId),
+          });
+        }
+      });
+      return nodes;
+    });
+  }
+
+  /**
+   * Perform an initial database synchronisation: get the k closest nodes
+   * from each seed node and add them to this database
+   * For now, we also attempt to establish a connection to each of them.
+   * If these nodes are offline, this will impose a performance penalty,
+   * so we should investigate performing this in the background if possible.
+   * Alternatively, we can also just add the nodes to our database without
+   * establishing connection.
+   * This has been removed from start() as there's a chicken-egg scenario
+   * where we require the NodeGraph instance to be created in order to get
+   * connections.
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public async syncNodeGraph() {
+    const backupSeedNodes = this.seedNodes;
```

Comment:
This reassignment of `this.seedNodes` can be changed to just checking whether the target node is in the seed node first and branching the logic in connecting to seed node vs connecting to non-seed node. This avoids this mutation of the `this.seedNodes` which can cause race conditions.

Reply:
Once this is changed, the `this.seedNodes` becomes a static property and doesn't change. Decentralised hoel punching may change this logic later in the future.

Reply:
And I need to change the async-init to bring in `ready` decorator that works while the class is also starting, this should make our `start()` more flexible and be able to call public methods. It will be however the programmer's responsibility to know when it is appropriate to call the public method, as it will depend on the `start` imperative order of side-effects.

Reply:
@tegefaulkes please resolve this when done.